src/main/java/us/kbase/sdk/common/TestLocalManager.java

```diff
@@ -0,0 +1,207 @@
+		public Path getRunDockerShFile() {
+			return testLocalDir.resolve(RUN_DOCKER_SH);
+		}
+
+		/** Get the path to the test local test config file.
+		 * @return the path.
+		 */
+		public Path getTestCfgFile() {
+			return testLocalDir.resolve(TEST_CFG);
+		}
+
+		/** Get whether the test config file was created as part of ensuring the test local
+		 * directory.
+		 * @return true if the test configuration file needed to be created. This means the user
+		 * will need to edit the file to add their credentials. 
+		 */
+		public boolean isCreatedTestCfgFile() {
+			return createdTestCfgFile;
+		}
+
+	}
+	
+	/** Create and / or restore the test local directory for an SDK app.
+	 * 
+	 * Will only overwrite files if they don't already exist.
+	 * 
+	 * @param moduleDir the SDK module root directory.
+	 * @param moduleName the name of the module.
+	 * @param dataVersion the version of the module's reference data, if applicable.
+	 * If provided, the "refdata" subfolder will be created if it doesn't exist. If is doesn't
+	 * exist, the test runner shell script will be regenerated to ensure it handles reference
+	 * data correctly.
+	 * @return information about the test local directory.
+	 * @throws IOException if a file or directory cannot be created.
+	 */
+	public static TestLocalInfo ensureTestLocal(
+			final Path moduleDir,
+			final String moduleName,
+			final Optional<String> dataVersion)
+			throws IOException {
+		final Path tlDir = requireNonNull(moduleDir, "moduleDir").resolve(TEST_LOCAL)
+				.toAbsolutePath();
+		final Path readmeFile = tlDir.resolve(README);
+		final Path testCfg = tlDir.resolve(TEST_CFG);
+		final Path runTestsSh = tlDir.resolve(RUN_TESTS_SH);
+		final Path runBashSh = tlDir.resolve(RUN_BASH_SH);
+		final Path runDockerSh = tlDir.resolve(RUN_DOCKER_SH);
+		final Map<String, Object> moduleContext = new HashMap<>();
+		// TODO CODE add a string checker that throws on whitespace only strings
+		moduleContext.put("module_name", requireNonNull(moduleName, "moduleName"));
+		Files.createDirectories(tlDir);
+		if (requireNonNull(dataVersion, "dataVersion").isPresent()) {
+			// TODO CODE add a string checker that throws on whitespace only strings
+			moduleContext.put("data_version", dataVersion.get());
+			final Path refDataDir = tlDir.resolve("refdata");
+			if (!Files.exists(refDataDir)) {
+				// We'll assume here that some bunghole hasn't created a file in the dir's place
+				// The missing refdata dir is treated as a signal that the user has added a
+				// refdata version to their kbase.yml file. That means the test runner
+				// needs to be regenerated to handle refdata.
+				// This all seems very clunky and could use a rethink
+				TemplateFormatter.formatTemplate(
+						"module_run_tests", moduleContext, runTestsSh.toFile()
+				);
+				Files.createDirectories(refDataDir);
+			}
+		}
+		boolean createdTestCfg = false;
+		// could make a template -> Path map here and DRY things up for all the but the last
```

Comment:
This would have a bit more impact (and one less superfluous "the") if worded like this:

> could make a template -> Path map here and DRY things up for all but the last if.
> Meh.

Reply:
Much better, fixed